### ISSUE TITLE
`@pure` decorator allowed for state-reading function

### ISSUE THREAD

**Issue by pcaversaccio** [vyperlang/vyper#3021]
### Version Information

* vyper Version (output of `vyper --version`): `0.3.4`
* OS: win
* Python Version (output of `python --version`): `3.9.2`

### What's your issue about?

I can successfully compile a contract with a function that reads from the current state the contract address (i.e. I use `self`) and use the decorator `@pure`. You can check the function [`_compute_address_self`](https://github.com/pcaversaccio/snekmate/blob/main/src/utils/Create2Address.vy#L19) and change the decorator to `@pure` instead (`@view` is the correct decorator here IMHO). The compilation will not fail. In my understanding (tbf my homebase is Solidity) this should fail.

### How can it be fixed?

Prevent state reading actions for `@pure` decorated functions.

**Comment by pkotzbach** [comment]
isn't function in your example calling another pure function?
https://vyper.readthedocs.io/en/stable/control-structures.html?highlight=pure#mutability
in docs it says that it is allowed

**Comment by charles-cooper** [comment]
it's ok for pure functions to call pure functions -- the issue here is that `self` is passed as a parameter, and `self` (which compiles to `ADDRESS`) is not pure.

**Comment by pcaversaccio** [comment]
yes exactly, as a comparison in Solidity this would fail:

```solidity
function computeAddress(bytes32 salt, bytes32 bytecodeHash) internal pure returns (address) {
    bytes32 data = keccak256(abi.encodePacked(bytes1(0xff), address(this), salt, bytecodeHash));
    return address(uint160(uint256(data)));
}
```

**Comment by charles-cooper** [comment]
fixed in #3027
